internal/globalMerge.ecl

```diff
@@ -0,0 +1,323 @@
+############################################################################## */
+IMPORT $.^.DBSCAN_Types AS Files;
+IMPORT ML_Core;
+IMPORT Std.system.Thorlib;
+
+/**
+  * globalMerge Module
+  *
+  * This module implements the stage 3 of the scalalbe and parallel DBSCAN algorithm [1].
+  * It merges the results from each node in stage 2 local clustering. It results in
+  * a synchronized cluster index for each data point.
+  * The cluster index is 1-based and the outliers are indexed as 0.
+  *
+  * Reference
+  * [1] Patwary, Mostofa Ali, et al. "A new scalable parallel DBSCAN algorithm using the
+  * disjoint-set data structure." Proceedings of the International Conference on High
+  * Performance Computing, Networking, Storage and Analysis. IEEE Computer Society Press, 2012.
+  */
+EXPORT globalMerge := MODULE
+
+  //Layout for Ultimate function and Loop_Func fuction
+  SHARED  l_ultimate := RECORD
+    UNSIGNED4 wi;
+    UNSIGNED4 id;
+    UNSIGNED4 parentID;        // the largest core point a data point belongs locally
+    UNSIGNED4 largestID := 0;  // the largest parentID of each data point among all local results
+    UNSIGNED4 ultimateID := 0; // the current cluster index
+  END;
+  //Ultimate function updates the largest parentID and the cluster index of each data point.
+  //It takes the dataset inl_utimate format and the number of data points in the dataset as input.
+  SHARED STREAMED DATASET(l_ultimate) ultimate(STREAMED DATASET(l_ultimate) dsin,
+                                                  UNSIGNED4 pointcount) := EMBED(C++:activity)
+    #include <stdio.h>
+    // The struct for each data point
+    struct upt
+    {
+      uint32_t wi;
+      uint32_t id;
+      uint32_t pid; //parentID
+    };
+
+    class MyStreamInlineDataset : public RtlCInterface, implements IRowStream
+      {
+        public:
+            MyStreamInlineDataset(IEngineRowAllocator * _resultAllocator,
+                                                IRowStream * _ds, uint32_t _pc)
+              :resultAllocator(_resultAllocator), ds(_ds), pc(_pc)
+              {
+                uptable = (upt*) rtlMalloc(pc * sizeof(upt));
+                for(uint32_t i = 0; i < pc; i++)
+                {
+                  uptable[i].wi = 0;
+                  uptable[i].id = 0;
+                  uptable[i].pid = 0;
+                };
+                calculated = false;
+                rc = 0;
+                lastgroupend = 0;
+                curWi = 0;
+                lastid = 0;
+              }
+              ~MyStreamInlineDataset(){
+                rtlFree(uptable);
+              }
+
+            RTLIMPLEMENT_IINTERFACE
+            //calculate the ultimate id
+            virtual const void *nextRow() override
+            {
+                if(!calculated){
+                    // read in the input dataset into a table
+                    while(true)
+                    {
```

Comment:
Also has a mixture of bracketing styles - sometimes on the same line, other times on the next line.

Reply:
Updated